### PR TITLE
API addition proposal + remove redundancy

### PR DESCRIPTION
**Pull request by remusao**
I played a bit with the idea of changing slightly the API #99 (while maintaining full backward compatibility) and I think I found some compromise.

- Add a `parse` method in the public API:
```js
> tldjs.parse('https://example.co.uk')
{ valid: true,
  hostname: 'foo.example.co.uk',
  suffix: 'co.uk',
  domain: 'example.co.uk',
  subdomain: 'foo' }
```
- All the other methods are still available and keep their semantic
- Add an additional option in `factory` to provide a custom method to extract a `hostname` from a `url` (by default it is `cleanHostValue`)
- Add a very fast hostname validity check in `cleanHostValue` to only pay the price of parsing if the argument is not already a valid hostname
- Change internal APIs to only accept a valid hostname + change arguments slightly to avoid re-computing public suffix, domain, etc. several times (`index.js` has some extra logic in parse to only extract the `hostname` once)

The result is a substantial increase in speed + no extra cost if the value given to any method of the public API is already a valid hostname.

Some timings (**if processing only clean hostnames**):
```js
While interpreting the results, keep in mind that each "op" reported by the benchmark is processing 24 domains
tldjs#isValid x 355,821 ops/sec ±0.70% (94 runs sampled)
tldjs#extractHostname x 333,040 ops/sec ±0.42% (96 runs sampled)
tldjs#tldExists x 128,460 ops/sec ±0.87% (96 runs sampled)
tldjs#getDomain x 41,242 ops/sec ±1.80% (87 runs sampled)
tldjs#getSubdomain x 41,626 ops/sec ±1.06% (95 runs sampled)
tldjs#getPublicSuffix x 40,460 ops/sec ±1.57% (92 runs sampled)
```
That is roughly *1M hostnames* processed per second (for `getDomain`, `getSubdomain`, `getPublicSuffix`).

Other timings (normal benchmark):
```js
While interpreting the results, keep in mind that each "op" reported by the benchmark is processing 24 domains
tldjs#isValid x 497,786 ops/sec ±0.55% (89 runs sampled)
tldjs#extractHostname x 52,050 ops/sec ±1.06% (85 runs sampled)
tldjs#tldExists x 36,349 ops/sec ±1.02% (92 runs sampled)
tldjs#getDomain x 18,385 ops/sec ±0.68% (91 runs sampled)
tldjs#getSubdomain x 18,335 ops/sec ±1.32% (96 runs sampled)
tldjs#getPublicSuffix x 18,289 ops/sec ±0.77% (94 runs sampled)
```
Which is about *430k hostnames* processed per second (but `extractHostname` is now the bottleneck).

If you think this is a reasonable change, we should close #100 (as it already implements the more strict `isValid` hostname validation), and I think I should stop optimizing the code :P It's probably more than fast enough now and if not, people can still plug their own fancy `extractHostname` function.

### CODE REVIEW

**Comment by ctavan** [comment]
@remusao I think this would be a very good improvement to the library!

Just one slight suggestion for more consistency: I think it should be `publicSuffix` instead of `suffix` in the result object of `tldjs.parse()`.

Only then will the result of parse strictly map to the current api:

```js
// tldjs.parse('https://example.co.uk')
{
  valid: true,  // tldjs.isValid()
  hostname: 'foo.example.co.uk',
  publicSuffix: 'co.uk', // tldjs.getPublicSuffix()
  domain: 'example.co.uk', // tldjs.getDomain()
  subdomain: 'foo',  // tldjs.getSubdomain()
}
```

**Comment by oncletom** [comment]
I support the `publicSuffix` thingy too.

Although I was wondering if it was sensible to have lazy properties instead, so as the computation would be done only on the used properties. Can be part of another pull request if it is relevant.

**Comment by remusao** [comment]
I just updated to rename `suffix` into `publicSuffix`. Regarding the lazy property, I actually tried it first, and the overhead of this was actually higher than just computing everything at once. So I'm not sure it's worth it. Actually I think it makes sense, it's not so much code so I guess V8 is able to optimize this like hell and inline a lot of it.

**Comment by oncletom** [comment]
> the overhead of this was actually higher than just computing everything at once

Well, that closes the case then 👍

File: test/tld.js [hunk 1]
```diff
@@ -56,7 +79,7 @@ describe('tld.js', function () {
       expect(tld.isValid('.com')).to.be(false);
     });
 
-    it('should be falsy on dotless hostname', function () {
+    it.skip('should be falsy on dotless hostname', function () {
```
Comment (oncletom) [inline test/tld.js]: The tests would not pass otherwise?

File: index.js [hunk 2]
```diff
@@ -1,17 +1,41 @@
...
+
+  if (valid) {
+    publicSuffix = getPublicSuffix(rules, hostname);
+    domain = getDomain(validHosts, publicSuffix, hostname);
+    subdomain = getSubdomain(hostname, domain);
+  }
+
+  return {
+    valid,
+    hostname,
```
Comment (oncletom) [inline index.js]: Thinking about it, the only property I would add would be `tldExists`, derived from the same method. Especially as now, thanks to you, we can retrieve a public suffix while having an unknown TLD.

What do you think?

Comment (remusao) [reply index.js]: Yes we can, this way we actually have an equivalent for each method of the API when using the `parse` function. I will add it.

File: test/tld.js (hunk 1 above)
Comment (remusao) [reply test/tld.js]: Actually, a hostname without dots seems to be valid (according to the firefox parser). So `localhost` is a valid hostname, strictly speaking, but we won't find a valid public suffix for it, so it will be rejected. Unless we specify it in `validHosts`. So I disabled these tests, should I remove them altogether?

We can double check if we have enough test cases to cover the behavior of all the functions for such corner cases (`getPublicSuffix`, `getDomain`, etc.) and make sure we offer a consistent behavior.

**Comment by remusao** [comment]
@oncletom I updated the PR with the following changes:

- Add `tldExists` as part of the output of `parse`
- Implement some kind of laziness in the implementation of `parse` to avoid paying the cost of all functions invocation. This works by having an optional `step` argument to the function, indicating if we want to stop early after a specific result has been computed (e.g.: `tldExists`, `publicSuffix`, etc.).

Current timings:
```javascript
tldjs#isValid x 524,277 ops/sec ±0.56% (92 runs sampled)
tldjs#extractHostname x 53,515 ops/sec ±0.25% (94 runs sampled)
tldjs#tldExists x 33,019 ops/sec ±0.65% (95 runs sampled)
tldjs#getPublicSuffix x 19,411 ops/sec ±1.32% (91 runs sampled)
tldjs#getDomain x 19,041 ops/sec ±0.95% (93 runs sampled)
tldjs#getSubdomain x 18,960 ops/sec ±0.61% (94 runs sampled)
tldjs#parse x 16,834 ops/sec ±1.21% (96 runs sampled)
```

**Comment by remusao** [comment]
Not sure what's going on with the installation on `node` in travis. Locally `testling` seems to be installed properly. Could this be a network error or something?

**Comment by oncletom** [comment]
Great work 👍 

> Not sure what's going on with the installation on node in travis. Locally testling seems to be installed properly. Could this be a network error or something?

I don't know, it seems to happen only with `node@8` 😞

**Comment by oncletom** [comment]
Issue seems to be related to Travis CI cache. Testling reports an issue which does not happen on the mocha side of things:

```
not ok 18 tld.js isValid method should detect valid hostname
  http://localhost:60900/__testling?show=true:11633:49
  callFn@http://localhost:60900/__testling/node_modules/mocha/mocha.js:4480:25
  run@http://localhost:60900/__testling/node_modules/mocha/mocha.js:4472:13
  runTest@http://localhost:60900/__testling/node_modules/mocha/mocha.js:4968:13
  http://localhost:60900/__testling/node_modules/mocha/mocha.js:5074:19
  next@http://localhost:60900/__testling/node_modules/mocha/mocha.js:4886:16
  http://localhost:60900/__testling/node_modules/mocha/mocha.js:4896:11
  next@http://localhost:60900/__testling/node_modules/mocha/mocha.js:4820:16
  http://localhost:60900/__testling/node_modules/mocha/mocha.js:4864:9
  timeslice@http://localhost:60900/__testling/node_modules/mocha/mocha.js:82:27
```

**Comment by oncletom** [comment]
And it seems to hang randomly on `node@8`. Always after the same test.

**Comment by remusao** [comment]
@oncletom Have you been able to reproduce the hanging? Is it while running `npm run test`? Or `testling`? Is it using `phantomjs` on Travis?

**Comment by oncletom** [comment]
I have not been able to reproduce the random hanging. The failing case with `node@4`.

Trying to identify the issue at the moment.

**Comment by oncletom** [comment]
It is related to the use of `'a'.repeat()`. I suspect tests are ran with PhantomJS, which does not have this ES method. Will check if Firefox can be used instead.

**Comment by remusao** [comment]
Ok, if this is the case, we can just replace the `repeat` with something else:
```javascript
function repeat(str, n) {
  var res = '';
  for (var i = 0; i < n; i += 1) {
    res += str;
  }
  return res;
}
```

**Comment by remusao** [comment]
I pushed it to give it a try, we can revert if it does not work.

**Comment by remusao** [comment]
Well, looks like it was that issue?

**Comment by oncletom** [comment]
Yeah and thinking about it it's weird, unless a different version of Phantom is in use depending on which version of Node is active…

**Comment by remusao** [comment]
Let's keep an eye on this! That's weird indeed... Thanks for merging.
